ethcore/src/service.rs

```diff
@@ -30,7 +30,7 @@ pub enum SyncMessage {
 		/// Hashes of blocks imported to blockchain
 		good: Vec<H256>,
 		/// Hashes of blocks not imported to blockchain
-		bad: Vec<H256>,
+		retracted: Vec<H256>,
```

Comment:
retracted != bad. Bad blocks are the ones which were not imported due to an error. You need additional field for dead blocks - ones which were removed from the canonical chain.